### PR TITLE
Explainer examples

### PR DESCRIPTION
**Pull request by otherdaniel**
This adds some examples to the explainer. I did my best to follow the current discussion, even though not all of the bits are finalized any may well change when actually specified.

### CODE REVIEW

File: explainer.md [hunk 1]
```diff
@@ -121,6 +121,106 @@ The currently proposed API differs in a number of aspects:
...
+
+The "safe" version will ignore configuration items that break its security
+guarantees:
+```js
+const an_unsafe_config = { 'allowElements': [ { name: 'script' } ] };
+element.setHTML("<script>", { sanitizer: an_unsafe_config });  // <div></div>
+element.setHTMLUnsafe("<script>", { sanitizer: an_unsafe_config });  // You now have a script. Congrats.
+```
+
+For elements, the HTML namespace is default. For attributes, the empty namespace.
```
Comment (annevk) [inline explainer.md:185]: null namespace. Empty string is normalized to null. Also applies below.

**Review by annevk — COMMENTED** [review]
Looks good, couple things that need some more attention.

File: explainer.md [hunk 2]
```diff
@@ -121,6 +121,106 @@ The currently proposed API differs in a number of aspects:
...
+```
+
+The "safe" methods remove all script-y content defined by the platform and
+let the rest pass:
+```js
+element.setHTML(`<a href=about:blank onclick=alert(1) onload=alert(2) id=myid class=something><script>alert(3);</script>`);
+// <div><a href="about:blank" id="myid" class="something"></a></div>
+```
+
+The operation of the built-in sanitizer can be configured to suit your
```
[the remark this answers is not in the document]
Comment (mozfreddyb) [reply explainer.md:172]: I don't think so. Sanitizer works here though?

[the remark this answers is not in the document]
Comment (annevk) [reply explainer.md:172]: It was meant as a general comment on this section.

File: explainer.md [hunk 3]
```diff
@@ -121,6 +121,106 @@ The currently proposed API differs in a number of aspects:
...
+
+Parsing observes its contexts:
+```js
+const example_tr = `<tr><td>A table row.</td></tr>`;
+const table = document.createElement("table");
+
+element.setHTML(example_tr);  // <div>A table row.</div>
+table.setHTML(example_tr);  // <table><tbody><tr><td>A table row.</td></tr></tbody></table>
+Document.parseHTML(example_tr);  // <html><body>A table row.</body></html>
+                                 // (Left out <head> for clarity.)
```
[the remark this answers is not in the document]
Comment (otherdaniel) [reply explainer.md]: Done.

File: explainer.md [hunk 4]
```diff
@@ -121,6 +121,106 @@ The currently proposed API differs in a number of aspects:
...
+All of these would have had identical results if the "unsafe" variants had
+been used.
+
+Parsing follows HTML parsing rules, unlike `innerHTML`, where it depends on the
+document type:
+```js
+const element_xml = new DOMParser().parseFromString("<html><body><div/></body></html>", "application/xhtml+xml").getElementsByTagName("div")[0];
+const example_not_xml = "<bLoCkQuOtE>bla";
+
+element_xml.getRootNode().contentType;  // 'application/xhtml+xml'
```
[the remark this answers is not in the document]
Comment (otherdaniel) [reply explainer.md]: Done.

File: explainer.md [hunk 5]
```diff
@@ -121,6 +121,106 @@ The currently proposed API differs in a number of aspects:
...
+element_xml.innerHTML = example_not_xml;  // Throws.
+element_xml.setHTML(example_not_xml);  // <div><blockquote>bla</blockquote></div>
+                                       // Note case and closing elements.
```
[the remark this answers is not in the document]
Comment (otherdaniel) [reply explainer.md]: Yes. When I simulate this by moving the context node into an HTML doc and then back in to the XML one, then I indeed get the namespace declaration. I fixed this by adding a namespace declaration to the original xhtml string (in DOMParser().parseFromString). That way, the xmlns decl is still in the output but on the surrounding `<div/>`, not on the `<blockquote>`.  Hope that makes sense...

File: explainer.md (hunk 1 above)
Comment (otherdaniel) [reply explainer.md]: Done.

File: explainer.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (annevk) [reply explainer.md]: That makes sense, but we still need to decide how to handle the corner case I suspect. Not sure what will fall out naturally when implementing this. Perhaps we should pretend the tag name is `<body>` when the namespace is not SVG, HTML, or MathML.

File: explainer.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (otherdaniel) [reply explainer.md:171]: No decision I know of. I'm happy to use any other term, or drop those examples for now.

This is easy to work around in the explanatory text. (E.g., "operation of the build-in sanitizer" -> "operation of setHTML and parseHTML methods".) But for the JS examples I can't really find a neutral way of expressing this. Is the note just above ("naming is preliminary") good enough, or should I put a different term here?

[the remark this answers is not in the document]
Comment (annevk) [reply explainer.md:171]: I had missed the note. I guess that works.

**Review by mozfreddyb — APPROVED** [review]

**Pull request merged**